### PR TITLE
[WIP] Updated the voting contract to support 4 voting types

### PR DESCRIPTION
**Pull request by robdmoore**
no snapshot, no weighting, weighting and partitioned weighting

### CODE REVIEW

File: src/dapp/src/features/vote-creation/RoundInfo.tsx [hunk 1]
```diff
@@ -15,6 +41,72 @@ const formSchema = zfd.formData({
   snapshotFile: zfd.text(z.string()),
   minimumVotes: zfd.numeric(z.number({ invalid_type_error: 'Should be a number' }).optional()),
 })
+//.superRefine(validateSnapshotCsv)
```
Comment (robdmoore) [inline src/dapp/src/features/vote-creation/RoundInfo.tsx:44]: todo: Can't get this working yet, OK to merge without though

File: src/dapp/src/features/vote-creation/RoundInfo.tsx [hunk 2]
```diff
@@ -71,7 +164,7 @@ export default function RoundInfo() {
                 field: 'snapshotFile',
                 hint: 'Upload snapshot .csv file',
                 longHint:
-                  'Upload a CSV file with a single column containing the addresses for the allowlist. Do not include a header line.',
+                  'Vote type = NO_SNAPSHOT: Upload empty file.\nVote type = NO_WEIGHTING: Upload a CSV file with a single column containing the addresses for the allowlist. Do not include a header line.\nOtherwise: Upload a CSV with 2 columns: address, weighting containing the addresses for the allowlist and their weighting. Do not include a header line.',
```
Comment (ricdikulous) [inline src/dapp/src/features/vote-creation/RoundInfo.tsx:167]: I think we should force them to have a header even in the case that it's a single column of addresses. Just to make the parsing consistent

File: src/dapp/src/shared/IPFSGateway.ts [hunk 3]
```diff
@@ -1,6 +1,14 @@
+enum VoteType {
```
Comment (ricdikulous) [inline src/dapp/src/shared/IPFSGateway.ts:1]: The `VoteType` enum has been defined in 3 separate places. Is that deliberate?

File: src/dapp/src/shared/csvSigner.ts [hunk 4]
```diff
@@ -7,24 +7,35 @@ interface SignedCsv {
...
   }
 
-  const results = Papa.parse(csv)
+  const results = Papa.parse<SnapshotRow>(csv, { header: true })
   const privateKey = ed.utils.randomPrivateKey()
   const publicKey = await ed.getPublicKeyAsync(privateKey)
   const signedCsv = await Promise.all(
     results.data
-      .filter((row) => !Array.isArray(row) || !!row[0])
+      .filter((row) => !Array.isArray(row) || !!row.address)
```
Comment (ricdikulous) [inline src/dapp/src/shared/csvSigner.ts:26]: Should only check the address here. Given we are using headers the `row` is an object where the headers are the keys

File: src/dapp/src/shared/csvSigner.ts [hunk 5]
```diff
@@ -7,24 +7,35 @@ interface SignedCsv {
...
       .map((row): Promise<SignedCsv> => {
         if (Array.isArray(row)) {
```
Comment (ricdikulous) [inline src/dapp/src/shared/csvSigner.ts:28]: This check will always fail given we set headers to true

File: src/dapp/src/shared/IPFSGateway.ts (hunk 3 above)
Comment (robdmoore) [reply src/dapp/src/shared/IPFSGateway.ts:1]: Deliberate, but I've added a comment to link them.

**Pull request merged**